### DISCUSSION TITLE
InvalidOperationException after upgrading to CSLA 8 prerelease 24010305

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#3647] [Questions]
@rockfordlhotka I have updated to the latest CSLA 8 prerelease, and am running into an issue that I can't figure out.  Outside of updating the Nuget packages to prerelease 24010305, I also made the following change to program.cs from this:

```c#
    builder.Services.AddCsla(options => options
        .AddAspNetCore()
        .AddServerSideBlazor()
        .PropertyChangedMode(Csla.ApplicationContext.PropertyChangedModes.Windows)
        .DataPortal(dpo =>
        {
            dpo.RegisterDataPortalCache(typeof(NdrnCache));
        })

```
to this:

```c#
    builder.Services.AddCsla(options => options
        .AddAspNetCore()
        .AddServerSideBlazor()
        .DataPortal(dpo =>
        {
            dpo.ClientSideDataPortal(cdpo => cdpo.DataPortalCacheType = typeof(NdrnCache));
        })

```
When I run the application, I'm prompted to my Identity Server and can log in, but when I am redirected back to my Blazor Server app I end up with the following exception:

```
Error: System.InvalidOperationException: Headers are read-only, response has already started.

   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowHeadersReadOnlyException()

   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseHeaders.Microsoft.AspNetCore.Http.IHeaderDictionary.set_SetCookie(StringValues value)

   at Microsoft.AspNetCore.Http.ResponseCookies.Append(String key, String value)

   at Csla.Blazor.State.SessionIdManager.GetSessionId()

   at Csla.Blazor.State.SessionManager.GetSession()

   at Csla.AspNetCore.Blazor.ApplicationContextManagerBlazor.GetClientContext(ExecutionLocations executionLocation)

   at Csla.ApplicationContext.get_ClientContext()

   at Csla.ApplicationContext.get_RuleSet()

   at Csla.Rules.BusinessRules.HasPermission(ApplicationContext applicationContext, AuthorizationActions action, Type objectType, Object[] criteria)

   at Csla.DataPortal`1.DoFetchAsync(Type objectType, Object criteria, Boolean isSync)

   at Csla.DataPortal`1.FetchAsync(Object[] criteria)

   at Dad.Business.LastClientsViewedInfoCollection.GetLastClientsViewedInfoCollectionAsync(Int32 employeeID, Int32 advocacyGroupID, IDataPortalFactory factory) in D:\Users\mtavares\Source\repos\Dad\Dad.Business\Client\LastClientsViewedInfoCollection.Designer.cs:line 74

   at Dad.Ui.Blazor.Shared.TopMenu.RefreshMenu(EntityTypeIDs entityType) in D:\Users\mtavares\Source\repos\Dad\Dad.Ui.Blazor\Shared\TopMenu.razor.cs:line 126

   at Dad.Ui.Blazor.Shared.TopMenu.OnInitializedAsync() in D:\Users\mtavares\Source\repos\Dad\Dad.Ui.Blazor\Shared\TopMenu.razor.cs:line 155

   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()

   at Dad.Ui.Blazor.Components.CustomComponentBase.InvokeWithBlazorServiceContext(Func`1 func) in D:\Users\mtavares\Source\repos\Dad\Dad.Ui.Blazor\Components\CustomComponentBase.cs:line 68
```

In wondering if the use of the cache had anything to do with the issue, I also removed that from the DataPortal options as well, but still ended up with the same issue.

I haven't upgraded to .NET 8 yet, and am running on .NET 7. Is there some other option that I need change in order to get this working in this CSLA version?

**Comment by rockfordlhotka** [comment]
The exception appears to be coming from the `SessionIdManager`, which is part of the new state management mechanism in Blazor.

The `SessionIdManager` creates and reads a cookie from the http request object off `HttpContext`.

What I get from the exception is that the id manager is attempting to append the cookie to the request object, and the request object is set to read-only by that time.

I think the solution is for your initial app landing page to inject an `ISessionManager` and call its `InitializeAsync` method in your landing page's `OnInitializedAsync`.

It _might_ be possible to do this in the `App.razor` page in .NET 7. I don't think that'll work in .NET 8, though that's an interesting thought.

The one trick on the landing page (at least if it is a server-static page) is that you can't interact with the data portal or the local/client context objects on that page - this is because the cookie isn't actually written to the browser until the page has finished rendering.

You can look at the [BlazorExample](https://github.com/MarimerLLC/csla/tree/main/Samples/BlazorExample) project to see how this is done.

I am thinking though, that perhaps the use of the state management should be opt-in. It _is_ opt-in on wasm, but not on the server, and maybe it should be opt-in on the server too. If someone doesn't use local/client context in their app, there should be no need to have this feature at all.

**Reply by mtavares628** [reply]
I tried your suggestion, but the `ISessionManager` doesn't have an `InitializeAsync` method.  I saw a `StateManager` class that does. However, I tried that in the `OnInitializedAsync` of my `App.razor`, but still got the same exception.

I'm not using server pre-rendering on my .NET 7 Blazor Server app at all, so I'm not sure if that matters or not. 

It would be great if we could make it opt-in, but if I'm using the wrong injected object, let me know.

**Reply by rockfordlhotka** [reply]
Sorry, you are right, it should be the StateManager service.

What happens is that any attempt to access local/client context (which the data portal does) in a Blazor app now gets those values from a per user store, identified by the ISessionIDManager service.

The default ID manager service uses a cookie on the HttpRequest.

I think the simplest solution for you is to use a different IContextManager. The Blazor default is the new behavior. The old behavior is the aspnetcore content manager.

Thinking of this, I think I may have done something to make this harder than it should be 🤔

I'm trying to drive out of a blizzard area in KS today, so can't provide a lot of detail.

Please look at the configuration class in CSLA dot Blazor project. In there I have code that removes all the IContextManager entries and adds the Blazor one. You'll need to do that in your program.cs file after the UseCsla call.

**Reply by mtavares628** [reply]
Thanks for the guidance.  I tried to use the aspnetcore context manager (`ApplicationContextManagerHttpContext`), but it was sporadic in that some of my pages loaded and some didn't.  My gut feeling is that it's because that context manager is only using `HttpContext` and not using the `AuthenticationStateProvider`.  When I looked back in the history, the config was never actually using that context manager, and it looks like you had just updated the existing `ApplicationContextManagerBlazor`.  So, I pulled down the copy of that file to before the state management changes and added it to my project:

```c#
    public class NdrnApplicationContextManagerBlazor : IContextManager, IDisposable
    {
        private ContextDictionary LocalContext { get; set; }
        private ContextDictionary ClientContext { get; set; }
        private IPrincipal CurrentPrincipal { get; set; }
        private readonly ClaimsPrincipal UnauthenticatedPrincipal = new();
        private bool disposedValue;

        /// <summary>
        /// Gets the current HttpContext instance.
        /// </summary>
        protected AuthenticationStateProvider AuthenticationStateProvider { get; private set; }

        /// <summary>
        /// Gets or sets a reference to the current ApplicationContext.
        /// </summary>
        public ApplicationContext ApplicationContext { get; set; }

        /// <summary>
        /// Gets the active circuit state.
        /// </summary>
        protected ActiveCircuitState ActiveCircuitState { get; private set; }

        /// <summary>
        /// Creates an instance of the object, initializing it
        /// with the required IServiceProvider.
        /// </summary>
        /// <param name="authenticationStateProvider">AuthenticationStateProvider service</param>
        /// <param name="activeCircuitState"></param>
        public NdrnApplicationContextManagerBlazor(AuthenticationStateProvider authenticationStateProvider, ActiveCircuitState activeCircuitState)
        {
            AuthenticationStateProvider = authenticationStateProvider;
            ActiveCircuitState = activeCircuitState;
            CurrentPrincipal = UnauthenticatedPrincipal;
            AuthenticationStateProvider.AuthenticationStateChanged += AuthenticationStateProvider_AuthenticationStateChanged;
            InitializeUser();
        }

        private void InitializeUser()
        {
            Task<AuthenticationState> task = default;
            try
            {
                task = AuthenticationStateProvider.GetAuthenticationStateAsync();
            }
            catch (InvalidOperationException ex)
            {
                task = Task.FromResult(new AuthenticationState((ClaimsPrincipal)UnauthenticatedPrincipal));
                string message = ex.Message;
                if (message.Contains(nameof(AuthenticationStateProvider.GetAuthenticationStateAsync))
                    && message.Contains(nameof(IHostEnvironmentAuthenticationStateProvider.SetAuthenticationState)))
                {
                    SetHostPrincipal(task);
                }
                else
                {
                    throw;
                }
            }
            AuthenticationStateProvider_AuthenticationStateChanged(task);
        }

        private void AuthenticationStateProvider_AuthenticationStateChanged(Task<AuthenticationState> task)
        {
            if (task is null)
            {
                CurrentPrincipal = UnauthenticatedPrincipal;
            }
            else
            {
                task.ContinueWith((t) =>
                {
                    if (task.IsCompletedSuccessfully && task.Result != null)
                        CurrentPrincipal = task.Result.User;
                    else
                        CurrentPrincipal = UnauthenticatedPrincipal;
                });
            }
        }

        /// <summary>
        /// Gets a value indicating whether this
        /// context manager is valid for use in
        /// the current environment.
        /// </summary>
        public bool IsValid
        {
            get { return ActiveCircuitState.CircuitExists; }
        }

        /// <summary>
        /// Gets a value indicating whether the context manager
        /// is stateful.
        /// </summary>
        public bool IsStatefulContext => true;

        /// <summary>
        /// Gets the current principal.
        /// </summary>
        public IPrincipal GetUser()
        {
            return CurrentPrincipal;
        }

        /// <summary>
        /// Attempts to set the current principal on the registered
        /// IHostEnvironmentAuthenticationStateProvider service.
        /// </summary>
        /// <param name="principal">Principal object.</param>
        public virtual void SetUser(IPrincipal principal)
        {
            if (!ReferenceEquals(CurrentPrincipal, principal))
            {
                if (principal is ClaimsPrincipal claimsPrincipal)
                {
                    CurrentPrincipal = principal;
                    SetHostPrincipal(Task.FromResult(new AuthenticationState(claimsPrincipal)));
                }
                else
                {
                    throw new ArgumentException("typeof(principal) != ClaimsPrincipal");
                }
            }
        }

        private void SetHostPrincipal(Task<AuthenticationState> task)
        {
            if (AuthenticationStateProvider is IHostEnvironmentAuthenticationStateProvider hostProvider)
                hostProvider.SetAuthenticationState(task);
        }

        /// <summary>
        /// Gets the local context.
        /// </summary>
        public ContextDictionary GetLocalContext()
        {
            if (LocalContext == null)
                LocalContext = new ContextDictionary();
            return LocalContext;
        }

        /// <summary>
        /// Sets the local context.
        /// </summary>
        /// <param name="localContext">Local context.</param>
        public void SetLocalContext(ContextDictionary localContext)
        {
            LocalContext = localContext;
        }

        /// <summary>
        /// Gets the client context.
        /// </summary>
        /// <param name="executionLocation"></param>
        public ContextDictionary GetClientContext(ApplicationContext.ExecutionLocations executionLocation)
        {
            if (ClientContext == null)
                ClientContext = new ContextDictionary();
            return ClientContext;
        }

        /// <summary>
        /// Sets the client context.
        /// </summary>
        /// <param name="clientContext">Client context.</param>
        /// <param name="executionLocation"></param>
        public void SetClientContext(ContextDictionary clientContext, ApplicationContext.ExecutionLocations executionLocation)
        {
            ClientContext = clientContext;
        }

        /// <summary>
        /// Dispose this object's resources.
        /// </summary>
        /// <param name="disposing"></param>
        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    AuthenticationStateProvider.AuthenticationStateChanged -= AuthenticationStateProvider_AuthenticationStateChanged;
                }
                disposedValue = true;
            }
        }

        /// <summary>
        /// Dispose this object's resources.
        /// </summary>
        public void Dispose()
        {
            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
    }
```
I then updated my program.cs as you suggested so it looks like this:

```c#
    builder.Services.AddCsla(options => options
        .AddAspNetCore()
        .AddServerSideBlazor()
        .DataPortal(dpo =>
        {
           dpo.ClientSideDataPortal(cdpo => cdpo.DataPortalCacheType = typeof(NdrnCache));
        }));

    var contextManagerType = typeof(Csla.Core.IContextManager);
    var managers = builder.Services.Where(i => i.ServiceType.Equals(contextManagerType)).ToList();
    foreach (var manager in managers)
        builder.Services.Remove(manager);
    builder.Services.AddScoped(typeof(IContextManager), typeof(NdrnApplicationContextManagerBlazor));

```
This seemed to get me in, but now my performance is extremely slow again, even with the cache implemented.  Additionally, I am seeing a bunch of `Exception thrown: 'System.InvalidOperationException' in Microsoft.AspNetCore.Components.Server.dll` messages in my Debug output window (not sure how to get more detail out of that window).  I feel like there is something else that I'm missing. 

Whenever you are safely out of your blizzard conditions, let me know if there is something else that I need to account for.  Thanks.

**Reply by mtavares628** [reply]
@rockfordlhotka - I'm still having performance issues when switching to this latest pre-release (8.0.0-R24031201).  I'm still on .NET 7, and don't have the performance issues with pre-release 8.0.0-R23122103.  Additionally, I'm still seeing a bunch of `Exception thrown: 'System.InvalidOperationException' in Microsoft.AspNetCore.Components.Server.dll` messages in the Debug output window and I'm not sure how to get more detail out of it to troubleshoot this.  Any thoughts would be appreciated.

**Reply by mtavares628** [reply]
@rockfordlhotka Upon further research, I think the errors are related to the new StateManager using TaskCompletionSource.  A google search turned up this info:

> The exception "System.InvalidOperationException" in Microsoft.AspNetCore.Components.Server.dll is thrown when the component is being statically rendered. This can happen when you're using a TaskCompletionSource and setting it to complete when it is already set to complete, or when you're doing some JsInterop calls on completed tasks.
> Here are some things you can try to fix this issue:
> Make sure that you're not using a TaskCompletionSource and setting it to complete when it is already set to complete.
> Make sure that you're not doing any JsInterop calls on completed tasks.

I'm not sure how to troubleshoot this further, since I'm not that familiar with this new StateManager.  Any guidance is appreciated.

**Reply by rockfordlhotka** [reply]
Logged as #3737

**Reply by rockfordlhotka** [reply]
@mtavares628 I did put the old Blazor 7 context manager back into CSLA to support pre-.NET 8 behavior - #3657

builder.Services.AddCsla(o=>o
    .AddAspNetCore()
    .AddServerSideBlazor(so => so.UseInMemoryApplicationContextManager = true));

**Reply by mtavares628** [reply]
@rockfordlhotka I had used that configuration in my testing of the latest pre-release:

```c#
    builder.Services.AddCsla(options => options
        .AddAspNetCore()
        .AddServerSideBlazor(bo =>
        {
            bo.UseInMemoryApplicationContextManager = true;
        })
        
        .DataPortal(dpo =>
        {
           dpo.ClientSideDataPortal(cdpo => cdpo.DataPortalCacheType = typeof(NdrnCache));
        })
        );
```

By using the the old Context manager, does it then bypass the State Manager altogether?

**Reply by rockfordlhotka** [reply]
Right - if you use the in memory context manager then you should not initialize or interact with the state management subsystem at all.
